### PR TITLE
Introduce public methods to access PlayIterator._host_states

### PR DESCRIPTION
**Pull request by mkrizek**
##### SUMMARY

Per https://github.com/ansible/ansible/pull/74175#issuecomment-824178733

ci_complete

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
`lib/ansible/executor/play_iterator.py`

### CODE REVIEW

File: lib/ansible/executor/play_iterator.py [hunk 1]
```diff
@@ -142,6 +151,14 @@ class PlayIterator:
     FAILED_RESCUE = 4
     FAILED_ALWAYS = 8
 
+    VALID_FAILED_STATES = [
+        FAILED_NONE,
+        FAILED_SETUP,
+        FAILED_TASKS,
+        FAILED_RESCUE,
+        FAILED_ALWAYS,
+    ]
```
Comment (Shrews) [inline lib/ansible/executor/play_iterator.py]: I think with moving to 3.8 on the controller, we can explore using some more advanced python features. This seems like a fantastic candidate for using an [enum class type](https://docs.python.org/3/library/enum.html).

File: lib/ansible/executor/play_iterator.py [hunk 2]
```diff
@@ -587,4 +588,19 @@ def _insert_tasks_into_state(self, state, task_list):
...
+        if not isinstance(state, HostState):
+            raise AnsibleAssertionError('Expected state to be a HostState but was a %s' % type(state))
+        self._host_states[hostname] = state
+
+    def set_run_state_for_host(self, hostname, run_state):
+        if not isinstance(run_state, IteratingStates):
+            raise AnsibleAssertionError('Expected run_state to be a IteratingStates but was %s' % (type(run_state)))
+        self._host_states[hostname].run_state = run_state
+
+    def set_fail_state_for_host(self, hostname, fail_state):
```
Comment (mkrizek) [inline lib/ansible/executor/play_iterator.py]: An alternative would be to have `clear_fail_state_for_host` that would set `fail_state` to `FailedState.NONE` and `set_fail_state_for_host` that would do `fail_state |= ...`. But it is not needed at the moment, so ¯\_(ツ)_/¯

File: lib/ansible/executor/play_iterator.py [hunk 3]
```diff
@@ -587,4 +588,19 @@ def _insert_tasks_into_state(self, state, task_list):
         return state
 
     def add_tasks(self, host, task_list):
-        self._host_states[host.name] = self._insert_tasks_into_state(self.get_host_state(host), task_list)
+        self.set_state_for_host(host.name, self._insert_tasks_into_state(self.get_host_state(host), task_list))
+
+    def set_state_for_host(self, hostname, state):
```
Comment (nitzmahone) [inline lib/ansible/executor/play_iterator.py]: Since this is new public API that's 3.8+, any reason we shouldn't type annotate these methods?

**Review by nitzmahone — APPROVED** [review]
Agreed, this is another long-needed change; also allows us to intercept/log these changes in the future for $reasons much easier than attrs did. Change looks good, just an open question if we want to add type annotations on the new public methods.

File: lib/ansible/executor/play_iterator.py (hunk 3 above)
Comment (mkrizek) [reply lib/ansible/executor/play_iterator.py]: +1, added.

**Review by nitzmahone — APPROVED** [review]

**Pull request merged**